### DISCUSSION TITLE
Why training an SCNN takes so much longer than its CNN counterpart？

### DISCUSSION THREAD

**Discussion by EliteSasuke** [jeshraghian/snntorch#125] [Q&A] [answered]
- snntorch: 0.4.10
- pytorch: 1.10.2+cuda11.3
- python: 3.7

Using SCNN and CNN with almost the same architecture to process the same data, the training time of the former(30m) is much longer than that of the latter(15s).
![image](https://user-images.githubusercontent.com/86088706/184659045-91991433-b0ed-4071-96f4-471f8ecea8a5.png)
Below is the structure of both:
```python
#SCNN
net = nn.Sequential(nn.Conv1d(1, 10, 10),
                    nn.MaxPool1d(5),
                    snn.Leaky(beta=beta, spike_grad=spike_grad, init_hidden=True),
                    nn.Conv1d(10, 20, 10),
                    nn.MaxPool1d(5),
                    snn.Leaky(beta=beta, spike_grad=spike_grad, init_hidden=True),
                    nn.Flatten(),
                    nn.Linear(100, 5),
                    snn.Leaky(beta=beta, spike_grad=spike_grad, init_hidden=True, output=True)
                    ).to(device)
```
```python
#CNN
net = nn.Sequential(nn.Conv1d(1, 10, 10),
                    nn.ReLU(inplace=True),
                    nn.MaxPool1d(5),
                    nn.Conv1d(10, 20, 10),
                    nn.ReLU(inplace=True),
                    nn.MaxPool1d(5),
                    nn.Flatten(),
                    nn.Linear(100, 5)
                    ).to(device)
```

**Comment by jeshraghian** [comment]
Can you share your training loop? 30m vs 15s doesn't sound right to me.

**Reply by EliteSasuke** [reply]
```python
#SCNN
spike_grad = surrogate.fast_sigmoid(slope=25)
beta = 0.95
num_steps = 200
batch_size=85
num_epochs=2
optimizer = torch.optim.Adam(net.parameters(), lr=5e-4, betas=(0.9, 0.999))
loss = nn.CrossEntropyLoss()
loader_train, loader_val, loader_test = get_dataloader(batch_size=batch_size)
counter=0

def forward_pass(net, data):
    mem_rec = []
    spk_rec = []
    utils.reset(net)  # resets hidden states for all LIF neurons in net
    data=torch.unsqueeze(data,1)
    for step in range(num_steps):
        spk_out, mem_out = net(data)
        spk_rec.append(spk_out)
        mem_rec.append(mem_out)
    return torch.stack(spk_rec), torch.stack(mem_rec)

for epoch in range(num_epochs):
    x = datetime.datetime.now()
    print(x.strftime("%X"))
    for i, (data, targets) in enumerate(iter(loader_train)):
        data = data.to(device)
        targets = targets.to(device)

        net.train()
        spk_rec,mem_rec = forward_pass(net, data)
        loss_val = torch.zeros((1), dtype=dtype, device=device)
        for step in range(num_steps):
            loss_val += loss(mem_rec[step], targets)

        # Gradient calculation + weight update
        optimizer.zero_grad()
        loss_val.backward()
        optimizer.step()

        # Store loss history for future plotting
        loss_hist.append(loss_val.item())

        acc = SF.accuracy_rate(spk_rec, targets)
        acc_hist.append(acc)
        if counter % 100 == 0:
             print(f"Epoch {epoch}, Iteration {i} \nTrain Loss: {loss_val.item():.2f}")
             print(f"Accuracy: {acc * 100:.2f}%\n")
        counter += 1
```
```python
#CNN
epochs = 2
learning_rate = 5e-4
batch_size = 85
train_loader, validation_loader, test_loader = get_dataloader(batch_size=batch_size)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(net.parameters(), lr=learning_rate,betas=(0.9, 0.999))
best_accuracy = 0

for epoch in range(0, epochs):
    x = datetime.datetime.now()
    print(x.strftime("%X"))
    print('Epoch: [%d]\t\t' % epoch, end='')
    net.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        data = torch.unsqueeze(data, 1)
        output = net(data.to(device))
        loss = criterion(output, target.to(device))
        loss.backward()
        optimizer.step()

    with torch.no_grad():
        net.eval()
        accuracy = test(net, test_loader)
        print('%2.2f%%' % accuracy)
        if accuracy > best_accuracy:
            torch.save(net.state_dict(), 'trained_model.pt')
            best_accuracy = accuracy
```

**Comment by jeshraghian** [comment]
The SCNN has 200 time steps, and so the forward pass will take 200ish times
longer. Generally, it makes more sense to compare run time on sequential
datasets of equal length (eg a recurrent CNN vs SCNN).

**Reply by EliteSasuke** [reply]
Thank you for your reply.
I adjusted the time step to 1（As before, the input data is not encoded as a spike train）, and the single-epoch training time of SCNN became 20s, which is still longer than CNN. Doesn't this contradict the theory that SNNs are more energy efficient?Or should I use other metrics instead of training time to verify the efficiency of SNN？

**Reply by dooseokjeong** [reply]
Energy efficiency of SNNs holds for forward passes when implemented in event-based neuromorphic hardware. The key is that the forward pass is based on asynchronous sparse events (spikes) so that such events are routed to target neurons (this process plus state variable updates is referred to as synaptic operation) only when the events occur unlike SNNs on GPUs, which need synchronous MAC operations even for no events. Regarding the difference in runtime between SCNN and CNN, I assume the runtime difference is due to the use of spiking units in place of ReLUs? If you want to make sure, please go through wallclock time breakdown line-by-line.

**Comment by jeshraghian** [comment, marked as answer]
Energy efficiency is with respect to running inference on neuromorphic
hardware, not gradient descent on GPUs. Efficiency training is a very
active field of research, and finding more efficient alternatives to
backprop is an important topic.

If youre looking for waya to quantify how your SCNN enhances efficiency
over CNN, Nengo-dl has a cool function that estimates the efficiency of SNN
models.

Providing similar metrics is on my to-do list with snntorch.

**Reply by EliteSasuke** [reply]
If I want to continue to do some research with snntorch (i.e. simulate SNN with python), where should I demonstrate the advantages of SNN?

**Comment by jeshraghian** [comment]
You can do this either experimentally or via simulation.

Experimental: port your model onto neuromorphic hardware (e.g., obtain
access to Loihi or SpinNaker) and use their power profiling tools.

Simulation: apply your model to the energy consumption estimation tools in
nengo-dl. Their library is well documented. This paper has an example of it
-
https://www.techrxiv.org/articles/preprint/Neuromorphic_Deep_Spiking_Neural_Networks_for_Seizure_Detection/20444970
